internal/trace/http_test.go

```diff
@@ -0,0 +1,778 @@
+				otelkey.String("http.user_agent", "foodownloader"),
+				otelkey.String("http.client_ip", "1.2.3.4"),
+			},
+		},
+		{
+			name:       "with http 1.1",
+			serverName: "my-server-name",
+			route:      "/user/:id",
+			method:     "GET",
+			requestURI: "/user/123",
+			proto:      "HTTP/1.1",
+			remoteAddr: "",
+			host:       "example.com",
+			url: &url.URL{
+				Path: "/user/123",
+			},
+			header: http.Header{
+				"User-Agent":      []string{"foodownloader"},
+				"X-Forwarded-For": []string{"1.2.3.4"},
+			},
+			tls: withTLS,
+			expected: []otelcore.KeyValue{
+				otelkey.String("http.method", "GET"),
+				otelkey.String("http.target", "/user/123"),
+				otelkey.String("http.scheme", "https"),
+				otelkey.String("http.flavor", "1.1"),
+				otelkey.String("http.server_name", "my-server-name"),
+				otelkey.String("http.route", "/user/:id"),
+				otelkey.String("http.host", "example.com"),
+				otelkey.String("http.user_agent", "foodownloader"),
+				otelkey.String("http.client_ip", "1.2.3.4"),
+			},
+		},
+		{
+			name:       "with http 2",
+			serverName: "my-server-name",
+			route:      "/user/:id",
+			method:     "GET",
+			requestURI: "/user/123",
+			proto:      "HTTP/2.0",
+			remoteAddr: "",
+			host:       "example.com",
+			url: &url.URL{
+				Path: "/user/123",
+			},
+			header: http.Header{
+				"User-Agent":      []string{"foodownloader"},
+				"X-Forwarded-For": []string{"1.2.3.4"},
+			},
+			tls: withTLS,
+			expected: []otelcore.KeyValue{
+				otelkey.String("http.method", "GET"),
+				otelkey.String("http.target", "/user/123"),
+				otelkey.String("http.scheme", "https"),
+				otelkey.String("http.flavor", "2"),
+				otelkey.String("http.server_name", "my-server-name"),
+				otelkey.String("http.route", "/user/:id"),
+				otelkey.String("http.host", "example.com"),
+				otelkey.String("http.user_agent", "foodownloader"),
+				otelkey.String("http.client_ip", "1.2.3.4"),
+			},
+		},
+	}
+	for idx, tc := range testcases {
+		r := testRequest(tc.method, tc.requestURI, tc.proto, tc.remoteAddr, tc.host, tc.url, tc.header, tc.tls)
+		got := HTTPServerAttributesFromHTTPRequest(tc.serverName, tc.route, r)
+		assertElementsMatch(t, tc.expected, got, "testcase %d - %s", idx, tc.name)
+	}
+}
+
+func TestHTTPAttributesFromHTTPStatusCode(t *testing.T) {
+	expected := []otelcore.KeyValue{
+		otelkey.Int("http.status_code", 404),
+		otelkey.String("http.status_text", "Not Found"),
+	}
+	got := HTTPAttributesFromHTTPStatusCode(http.StatusNotFound)
+	assertElementsMatch(t, expected, got, "with valid HTTP status code")
+	assert.ElementsMatch(t, expected, got)
+	expected = []otelcore.KeyValue{
+		otelkey.Int("http.status_code", 499),
+	}
+	got = HTTPAttributesFromHTTPStatusCode(499)
+	assertElementsMatch(t, expected, got, "with invalid HTTP status code")
+}
+
+func TestSpanStatusFromHTTPStatusCode(t *testing.T) {
+	for code := 0; code < 1000; code++ {
+		expected := getExpectedGRPCCodeForHTTPCode(code)
```

Comment:
Seems a bit like we are testing with a copy of the functional code which I'm not sure is that helpful other than it being a change indicator. :shrug:

Reply:
Yeah, I had no idea how to test it without copying the code outright. I was rather disinclined to do the `map[http.StatusCode]codes.Code` stuff. Too boring. :)

Reply:
fair enough.